src/libs/ascent/runtimes/ascent_main_runtime.cpp

```diff
@@ -616,24 +615,77 @@ AscentRuntime::EnsureDomainIds()
     }
 
     int domain_offset = 0;
+    int max_num_domains = num_domains;
+
 #ifdef ASCENT_MPI_ENABLED
     int *domains_per_rank = new int[comm_size];
     MPI_Allgather(&num_domains, 1, MPI_INT, domains_per_rank, 1, MPI_INT, mpi_comm);
     for(int i = 0; i < m_rank; ++i)
     {
       domain_offset += domains_per_rank[i];
     }
+    for(int i = 0; i < comm_size; i++)
+    {
+      if(domains_per_rank[i] > max_num_domains)
+        max_num_domains = domains_per_rank[i];
+    }
     delete[] domains_per_rank;
 #endif
+
+    std::unordered_set<int> local_unique_ids;
+
+    int *domain_ids = new int[max_num_domains];
     for(int i = 0; i < num_domains; ++i)
     {
       conduit::Node &dom = m_source.child(i);
 
       if(!dom.has_path("state/domain_id"))
       {
          dom["state/domain_id"] = domain_offset + i;
+         domain_ids[i] = domain_offset + i;
+         local_unique_ids.insert(domain_ids[i]);
       }
+      else
+      {
+         domain_ids[i] = dom["state/domain_id"].to_int32();
+         local_unique_ids.insert(domain_ids[i]);
+      }
+    }
+
+    if(local_unique_ids.size() != num_domains)
+    {
+      ASCENT_ERROR("Local Domain IDs are not unique ");
```

Comment:
In the mpi case, this would hang if one mpi task has bad local ids. (before getting deeper into mpi logic)

We need to add mpi error checking (flag all is ok, etc)